File: src/main/java/me/hydos/mowingsim/engine/io/Screen.java
package me.hydos.mowingsim.engine.io;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;
import me.hydos.mowingsim.engine.Rosella;

import java.util.List;

import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.vulkan.VK10.vkEnumerateInstanceExtensionProperties;

public class Screen {

	public final String title;
	public final int width, height;
	private final long windowPtr;
	private final List<MainLoopCallback> callbacks = new ObjectArrayList<>();

	public Screen(String title, int width, int height, boolean windowResizable) {
		this.title = title;
		this.width = width;
		this.height = height;
		glfwInit();

		windowPtr = glfwCreateWindow(width, height, title, 0, 0);
		glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);
		glfwWindowHint(GLFW_RESIZABLE, windowResizable ? GLFW_TRUE : GLFW_FALSE);

		Runtime.getRuntime().addShutdownHook(new Thread(() -> {
			glfwDestroyWindow(windowPtr);
			glfwTerminate();
		}));
	}

	public void start(Rosella engine) {
		while (!glfwWindowShouldClose(windowPtr)) {
			glfwPollEvents();
			for (MainLoopCallback callback : callbacks) {
				callback.run();
			}
		}
	}

	public void onMainLoop(MainLoopCallback callback) {
		callbacks.add(callback);
	}

	public interface MainLoopCallback {
		void run();
	}
}

File: src/main/kotlin/me/hydos/rosella/io/Screen.kt
package me.hydos.rosella.io

import it.unimi.dsi.fastutil.objects.ObjectArrayList
import me.hydos.rosella.Rosella
import org.lwjgl.glfw.GLFW.*

class Screen(title: String, width: Int, height: Int, windowResizable: Boolean = true) {
	internal val windowPtr: Long
	private val loopCallbacks: MutableList<() -> Unit> = ObjectArrayList()
	private val closeCallbacks: MutableList<() -> Unit> = ObjectArrayList()

	fun start(engine: Rosella) {
		glfwSetFramebufferSizeCallback(windowPtr, engine::windowResizeCallback)

		while (!glfwWindowShouldClose(windowPtr)) {
			glfwPollEvents()

			for (callback in loopCallbacks) {
				callback()
			}
		}
	}

	fun onMainLoop(callback: () -> Unit) {
		loopCallbacks.add(callback)
	}

	fun onWindowClose(function: () -> Unit) {
		closeCallbacks.add(function)
	}

	init {
		if (!glfwInit()) {
			throw RuntimeException("Cannot Initialize GLFW")
		}
		glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API)
		glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE)
		glfwWindowHint(GLFW_RESIZABLE, windowResizable.asGlfw())
		windowPtr = glfwCreateWindow(width, height, title, 0, 0)

		Runtime.getRuntime().addShutdownHook(Thread {
			for (callback in closeCallbacks) {
				callback()
			}
			glfwDestroyWindow(windowPtr)
			glfwTerminate()
		})
	}
}

private fun Boolean.asGlfw(): Int = if (this) GLFW_TRUE else GLFW_FALSE
